src/layers/rotary_embedding_chatglm2.cpp

```diff
@@ -111,41 +113,87 @@ void ChatGLM2RotaryEmbedding::forward(float *buf, int bufStride, int batch_size,
                     p1[i] = p1[i] * pcos[i] - p1[i + 1] * psin[i];
                     p1[i + 1] = p1[i + 1] * pcos[i] + t1 * psin[i];
                 }
-                off += bufStride;
+                off += qStride;
             }
         }
     }
 }
 
+inline void ChatGLM2RotaryEmbedding::prepare_sincos(__m512 a, __m512 b, __m512 *result) {
+    const __m512i mask = _mm512_set_epi32(
+            0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10, 0x0e, 0x0c, 0x0a, 0x08, 0x06, 0x04, 0x02, 0x00);
+
+    *result = _mm512_permutex2var_ps(a, mask, b);
+}
+
+inline void ChatGLM2RotaryEmbedding::interleave_qk(__m512 a, __m512 b, __m512 *result0, __m512 *result1) {
+    const __m512i mask0 = _mm512_set_epi32(
+            0x1e, 0x1c, 0x1a, 0x18, 0x16, 0x14, 0x12, 0x10, 0x0e, 0x0c, 0x0a, 0x08, 0x06, 0x04, 0x02, 0x00);
+
+    const __m512i mask1 = _mm512_set_epi32(
+            0x1f, 0x1d, 0x1b, 0x19, 0x17, 0x15, 0x13, 0x11, 0x0f, 0x0d, 0x0b, 0x09, 0x07, 0x05, 0x03, 0x01);
+
+    *result0 = _mm512_permutex2var_ps(a, mask0, b);
+    *result1 = _mm512_permutex2var_ps(a, mask1, b);
+}
+
+inline void ChatGLM2RotaryEmbedding::deinterleave_qk(__m512 a, __m512 b, __m512 *result0, __m512 *result1) {
+    const __m512i mask0 = _mm512_set_epi32(
+            0x17, 0x07, 0x16, 0x06, 0x15, 0x05, 0x14, 0x04, 0x13, 0x03, 0x12, 0x02, 0x11, 0x01, 0x10, 0x00);
+
+    const __m512i mask1 = _mm512_set_epi32(
+            0x1f, 0x0f, 0x1e, 0x0e, 0x1d, 0x0d, 0x1c, 0x0c, 0x1b, 0x0b, 0x1a, 0x0a, 0x19, 0x09, 0x18, 0x08);
+
+    *result0 = _mm512_permutex2var_ps(a, mask0, b);
+    *result1 = _mm512_permutex2var_ps(a, mask1, b);
+}
+
 void ChatGLM2RotaryEmbedding::forward(
-        float *query, float *key, int qStride, int kStride, const int *qk_shape, const int *position_ids) {
+        bfloat16_t *query, bfloat16_t *key, int qStride, int kStride, const int *qk_shape, const int *position_ids) {
     int dim = inv_freq_size * 2;
     REQUIRES(dim == qk_shape[3], "Incorrect shape, last dimention is not the head size.");
     const int batch_size = qk_shape[0];
     const int seq_len = qk_shape[1];
-    const int head_num = qk_shape[2] +  qk_shape[4];
+    const int head_num = qk_shape[2] + qk_shape[4];
     const int half = inv_freq_size;
-
-#pragma omp parallel for
+    // #pragma omp parallel for
```

Comment:
Why not to use it to parallel for?

Reply:
for the same parameters：
2 sockets:
without  #pragma omp parallel for
Througput:      12.+ tokens/s

with #pragma omp parallel for
Througput:      11.+ tokens/s

we can assume that no benefit gets from OMP PARALLEL.

Reply:
maybe we are better to align with other rotary embedding implement.